pingcap/tidb-operator
# Support BR on Azure Blob Storage

### What problem does this PR solve?

Closes #4533

### What is changed and how does it work?

- adding `azblob` field in CR -`Backup`/`Restore`/`ScheduleBackup` to control backup or restore on  Azure Blob Storage.
- adding credentials for provider azblob using AD or shared key.
- adding `envFrom` field in `TidbCluster`-`TiKVSpec`/`PDSpec`/... - `ComponentSpec` for using AAD credentials on any cloud or on-premises.
- adding GC for azblob while `cleanPolicy` field in CR-`Backup` is Delete.

### Code changes

- [x] Has examples related yaml change

### Tests



Manual test (add detailed scripts or steps below):

- `Backup` CR
  - generate some data in the database
  
  - > Using shared key credentials 
    
    1. run `kubectl create secret generic azblob-secret --from-literal=AZURE_STORAGE_ACCOUNT={test_account} --from-literal=AZURE_STORAGE_KEY={test_key}`
    2. run `kubectl apply -f examples/backup/backup-azblob.yaml`
    
  - > Using AAD credentials 
 
    1. run `kubectl create secret generic azblob-secret --from-literal=AZURE_CLIENT_ID={test_cid} --from-literal=AZURE_TENANT_ID={test_tid} --from-literal=AZURE_CLIENT_SECRET={test_secret} --from-literal=AZURE_STORAGE_ACCOUNT={test_account}`
    2. run `kubectl patch tc basic --type merge -p '{"spec":{"tikv":{"envFrom":[{"secretRef":{"name":"azblob-secret"}}]}}}'`, then wait for TiKV to finish restarting
    3. look forward to setting `sendCredToTikv` value to false, before running `kubectl apply -f examples/backup/backup-azblob.yaml`
    
  - wait for the Job to complete, then the backup metadata and sst files will appear in the /test1/t1 folder on Azure Blob Storage
  - run `kubectl delete -f examples/backup/backup-azblob.yaml`  and notice that backup files are deleted because `spec.cleanPolicy` is set to Delete
 
- `Restore` CR
  - delete data from the database  
  - run `kubectl apply -f examples/backup/restore-azblob.yaml`. the location should point to the same azblob folder in the `Backup` CR     
  - wait for the Job to complete, then the metadata and sst files will be restored into the database

- `BackupSchedule` CR
  - generate some data in the database
  - Refer to `Backup` CR above in terms of using shared key or AAD credentials
  - run `kubectl apply -f examples/backup/backup-schedule-azblob.yaml`, which schedules a backup job every minute.
  - wait for a few minutes and there should be multiple backups in /test1/t1 folder on azblob
  - run `kubectl delete -f examples/backup/backup-schedule-azblob.yaml` and notice that backup files are deleted because `spec.backupTemplate.cleanPolicy` is set to Delete

### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support BR on Azure Blob Storage in TiDB Operator
```
